src/msgs/measurements.rs

```diff
@@ -281,49 +413,470 @@ bitflags! {
+        }
+
+        Ok(blocks)
+    }
+}
+
+#[derive(Debug)]
+pub enum ParseMeasurementsError {
+    MaxSlotNumberExceeded,
+    MaxBlocksExceeded,
+    MaxOpaqueDataSizeExceeded,
+    Read(ReadError),
+    Blocks(ParseMeasurementBlocksError),
+    ContentChanged(ParseContentChangedError),
+    Opaque(ParseOpaqueDataError),
+}
+
+impl From<ReadError> for ParseMeasurementsError {
+    fn from(e: ReadError) -> Self {
+        ParseMeasurementsError::Read(e)
+    }
+}
+
+impl From<ParseMeasurementBlocksError> for ParseMeasurementsError {
+    fn from(e: ParseMeasurementBlocksError) -> Self {
+        ParseMeasurementsError::Blocks(e)
+    }
+}
+
+impl From<ParseContentChangedError> for ParseMeasurementsError {
+    fn from(e: ParseContentChangedError) -> Self {
+        ParseMeasurementsError::ContentChanged(e)
+    }
+}
+
+impl From<ParseOpaqueDataError> for ParseMeasurementsError {
+    fn from(e: ParseOpaqueDataError) -> Self {
+        ParseMeasurementsError::Opaque(e)
+    }
+}
+
 #[derive(Debug, Clone, PartialEq)]
 pub struct Measurements {
     // Param1
-    pub num_measurement_indices: u8,
+    pub total_blocks: u8,
 
-    // Param2
+    // Bits [5:4] of Param2
     pub content_changed: ContentChanged,
 
     // slot_id of cert chain specified in GET_MEASUREMENTS request
     // Only used for signature generation.
     // Set to 0 if no signature in this message
+    //
+    // Bits 3:0 of Param2
     pub slot_id: u8,
 
-    pub num_blocks: u8,
-    pub blocks: [MeasurementBlock; config::MAX_MEASUREMENT_BLOCKS],
-
+    pub blocks: MeasurementBlocks,
     pub nonce: Nonce,
     pub opaque_data: OpaqueData,
-    pub signature: SignatureBuf,
+    pub signature: Option<SignatureBuf>,
+}
+
+impl Measurements {
+    /// Return a Measurements message filled in with the total number of
+    ///  measurement blocks available.
+    ///
+    /// The value created by this function is returned when `Param2` of
+    /// `GetMeasurements` is set to `0x0`.
+    pub fn new_with_num_measurements(
+        total_blocks: u8,
+        content_changed: ContentChanged,
+        slot_id: u8,
+        nonce: Nonce,
+        opaque_data: OpaqueData,
+        signature: Option<SignatureBuf>,
+    ) -> Measurements {
+        Measurements {
+            total_blocks,
+            content_changed,
+            slot_id,
+            blocks: MeasurementBlocks::default(),
+            nonce,
+            opaque_data,
+            signature,
+        }
+    }
+
+    /// Return a measurements message with all measurements filled in or a
+    /// single measurement block, depending upon the `GetMeasurements` request.
+    pub fn new_with_measurements(
+        content_changed: ContentChanged,
+        slot_id: u8,
+        blocks: MeasurementBlocks,
+        nonce: Nonce,
+        opaque_data: OpaqueData,
+        signature: Option<SignatureBuf>,
+    ) -> Measurements {
+        Measurements {
+            total_blocks: 0,
```

Comment:
This is definitely an SPDM peculiarity. `total_blocks` is only filled in when the user requests the total number of measurement indices (blocks) available in the request, and not the actual measurements. It's "reserved" otherwise, which to me implies setting the value to 0. That's what happens in most other cases of reserved so I stuck to the convention.

> When Param2 in the requested measurement operation
is 0 , this parameter shall return the total number of
measurement indices on the device. Otherwise, this field
is reserved.